plugins/catalog-backend/src/ingestion/LocationReaders.ts

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { getVoidLogger } from '@backstage/backend-common';
+import {
+  EntityPolicies,
+  EntityPolicy,
+  LocationSpec,
+} from '@backstage/catalog-model';
+import { Logger } from 'winston';
+import { AnnotateLocationEntityProcessor } from './processors/AnnotateLocationEntityProcessor';
+import { EntityPolicyProcessor } from './processors/EntityPolicyProcessor';
+import { FileReaderProcessor } from './processors/FileReaderProcessor';
+import { GithubReaderProcessor } from './processors/GithubReaderProcessor';
+import * as result from './processors/results';
+import {
+  LocationProcessor,
+  LocationProcessorDataResult,
+  LocationProcessorEntityResult,
+  LocationProcessorErrorResult,
+  LocationProcessorLocationResult,
+  LocationProcessorResult,
+  LocationProcessorSink,
+} from './processors/types';
+import { YamlProcessor } from './processors/YamlProcessor';
+import { LocationReader, ReadLocationResult } from './types';
+
+// The max amount of nesting depth of generated work items
+const MAX_DEPTH = 10;
+
+/**
+ * Implements the reading of a location through a series of processor tasks.
+ */
+export class LocationReaders implements LocationReader {
+  private readonly logger: Logger;
+  private readonly processors: LocationProcessor[];
+
+  static defaultProcessors(
+    entityPolicy: EntityPolicy = new EntityPolicies(),
+  ): LocationProcessor[] {
+    return [
+      new FileReaderProcessor(),
+      new GithubReaderProcessor(),
+      new YamlProcessor(),
+      new EntityPolicyProcessor(entityPolicy),
+      new AnnotateLocationEntityProcessor(),
+    ];
+  }
+
+  constructor(
+    logger: Logger = getVoidLogger(),
+    processors: LocationProcessor[] = LocationReaders.defaultProcessors(),
+  ) {
+    this.logger = logger;
+    this.processors = processors;
+  }
+
+  async read(location: LocationSpec): Promise<ReadLocationResult> {
+    const output: ReadLocationResult = { entities: [], errors: [] };
+    let items: LocationProcessorResult[] = [result.location(location, false)];
+
+    for (let depth = 0; depth < MAX_DEPTH; ++depth) {
+      const newItems: LocationProcessorResult[] = [];
+      const sink: LocationProcessorSink = i => newItems.push(i);
```

Comment:
Could call it `emit` here too, and possibly have the type called `LocationProcessorEmit`? Kinda weird that it switches name all of a sudden without really switching context much.

Reply:
Agreed, that was sloppy naming.